### PR TITLE
[improvement] improve install.sh if then statement

### PR DESCRIPTION
**Pull request by tracehh**
## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

### CODE REVIEW

**Comment by tracehh** [comment]
see issue: https://github.com/apache/dolphinscheduler/issues/6781

**Comment by zhongjiajie** [comment]
Hi @tracehh , to target specific issue from PR, we recommend use keyword `fix: #issue_id` or `close: #issue_id` or `closes: #issue_id` in your PR describe. I would not only connect issue to PR but also close issue automatically when PR is be closed.

**Comment by zhongjiajie** [comment]
> Hi @tracehh , to target specific issue from PR, we recommend use keyword `fix: #issue_id` or `close: #issue_id` or `closes: #issue_id` in your PR describe. I would not only connect issue to PR but also close issue automatically when PR is be closed.

Hi, @tracehh could you please target issue with specific syntax?

File: install.sh [hunk 1]
```diff
@@ -36,46 +45,95 @@ if [ $dbtype == "postgresql" ];then
...
 sed -i ${txt} "s@^#*server.port=.*@server.port=${apiServerPort}@g" conf/application-api.properties
 
 # Change configuration in conf/registry.properties
+registryServers=$(charEscape ${registryServers})
+registryPluginDir=$(charEscape ${registryPluginDir})
+registryPluginName=$(charEscape ${registryPluginName})
+sed -i ${txt} "s@^#*registry.servers=.*@registry.servers=${registryServers}@g" conf/registry.properties
 sed -i ${txt} "s@^#*registry.plugin.dir=.*@registry.plugin.dir=${installPath}/${registryPluginDir}@g" conf/registry.properties
 sed -i ${txt} "s@^#*registry.plugin.name=.*@registry.plugin.name=${registryPluginName}@g" conf/registry.properties
-sed -i ${txt} "s@^#*registry.servers=.*@registry.servers=${registryServers}@g" conf/registry.properties
```
Comment (zhongjiajie) [inline install.sh]: Why does we delete this line?

File: install.sh [hunk 2]
```diff
@@ -88,26 +146,21 @@ fi
 # 3.scp resources
 echo "3.scp resources"
 sh ${workDir}/script/scp-hosts.sh
-if [ $? -eq 0 ]
-then
+if [ $? -eq 0 ];then
```
Comment (zhongjiajie) [inline install.sh:35]: Good catch!

**Review by zhongjiajie — COMMENTED** [review]
Hi @tracehh Code look good to me. but did you test in you local env? shell script need to test carefully could we do not have unittest to it

File: install.sh (hunk 1 above)
Comment (zhongjiajie) [reply install.sh]: I got your point, you just move this line to the top of this group. Please recover it to the tail, cause best practices of pull requests is only change code that have to change. @tracehh

Comment (tracehh) [reply install.sh]: It is redundant

**Comment by zhongjiajie** [comment]
It seem function do not work in my env. I use ]

* OS: mac
* Bash: GNU bash, version 5.1.8(1)-release (x86_64-apple-darwin19.6.0)

```shell
function charEscape(){
    inputParam=$1
    echo ${inputParam}
    if [[ $1 =~ '@' ]]; then
        tmp=$(echo inputParam | sed "s/@/\\\\@/g")
        inputParam=${tmp}
    fi
    echo "$inputParam"
}

input="abc@123"
input=$(charEscape ${input})
echo ${input}
```

And  the result is `inputParam`

File: install.sh [hunk 3]
```diff
@@ -21,6 +21,15 @@ workDir=`cd ${workDir};pwd`
 
 source ${workDir}/conf/config/install_config.conf
 
+charEscape(){
+    inputParam=$1
+    if [[ $1 =~ '@' ]]; then
+        tmp=$(echo inputParam | sed "s/@/\\\\@/g")
```
Comment (zhongjiajie) [inline install.sh]: It seem we miss shell variable syntax here @tracehh 
```suggestion
        tmp=$(echo ${inputParam} | sed "s/@/\\\\@/g")
```

but the most important thing is, could we reduce the complexity of function? maybe just

```shell
function charEscape(){
  echo ${1//@/\\\\@} 
}
```

according to https://stackoverflow.com/a/2871187/7152658, and this work on my env, with 

* OS: mac
* Bash: GNU bash, version 5.1.8(1)-release (x86_64-apple-darwin19.6.0)

Comment (zhongjiajie) [reply install.sh]: WDYT  @tracehh

Comment (tracehh) [reply install.sh]: Year, missing {} for inputParam.

Comment (tracehh) [reply install.sh]: Nice advice, simplify it by reducing if

Comment (zhongjiajie) [reply install.sh]: >  Nice advice, simplify it by reducing if

Maybe you should test wether it work or not in your env.

File: install.sh (hunk 1 above)
Comment (zhongjiajie) [reply install.sh]: > It is redundant

I do not think so, you just move this line from L78 to L134, we could see screenshot as below.

image

Comment (zhongjiajie) [reply install.sh]: > It is redundant

I do not think so, you just move this line from L78 to L134, we could see screenshot as below.

image

**Comment by tracehh** [comment]
sed is removed in install.sh file. Another way to config them now? @zhongjiajie

**Comment by zhongjiajie** [comment]
It seem that remove in https://github.com/apache/dolphinscheduler/pull/6840. I going to check what happen to it.

**Comment by zhongjiajie** [comment]
> It seem that remove in #6840. I going to check what happen to it.

Hi @tracehh , It seem that this PR should be close, could our remote code delete and we would not use sed to change configure file anymore. Maybe you should found another issue in #5689 or just search issue in our issue list.

**Comment by tracehh** [comment]
1bbfc3c1cd99709ba6d587e42ce7f32
when I find out this issue, I think there is a better way to replace 'sed'

**Comment by tracehh** [comment]
As 'sed' removed in install.sh, there are some format style, I change them, please look look @zhongjiajie

File: install.sh [hunk 4]
```diff
@@ -32,26 +32,21 @@ fi
...
 # 4.delete zk node
 echo "4.delete zk node"
-
 sh ${workDir}/script/remove-zk-node.sh $zkRoot
 
-
 # 5.startup
 echo "5.startup"
-sh ${workDir}/script/start-all.sh
+sh ${workDir}/script/start-all.sh
```
Comment (zhongjiajie) [inline install.sh]: Please keep file end with new line 
```suggestion
sh ${workDir}/script/start-all.sh

```

**Review by zhongjiajie — COMMENTED** [review]
Change look good to me, but please make sure we have new line in the end of file

**Comment by zhongjiajie** [comment]
@tracehh BTW please change PR title for this PR, and remember close #6781  cause we do not fix this issue anymore.

File: install.sh (hunk 4 above)
Comment (zhongjiajie) [reply install.sh]: @tracehh could you have time to add new line in the end of this file?

**Review by zhongjiajie — APPROVED** [review]
LGTM, thanks. You put a lot of effort into this feature, and although it ended up being just a code style change, you created two PRs earlier to try to fix the `sed` problem in `shell` script, and in the end didn't think we'd just remove it. 👍

**Review by CalvinKirs — APPROVED** [review]
+1

**Comment by kezhenxu94** [comment]
Thank you all the same. 👍 @tracehh

**Pull request merged**